certora/specs/Transfer.spec

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+methods {
+    function libSafeTransfer(address, address, uint256) external envfree;
+    function libSafeTransferFrom(address, address, address, uint256) external envfree;
+    function balanceOf(address, address) external returns (uint256) envfree;
+    function allowance(address, address, address) external returns (uint256) envfree;
+    function totalSupply(address) external returns (uint256) envfree;
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.allowance(address, address) external => DISPATCHER(true);
+    function _.totalSupply() external => DISPATCHER(true);
+}
+
+ghost mapping(address => mathint) myBalances
+{
+    init_state axiom (forall address token. myBalances[token] == 0);
+}
+
+function summarySafeTransferFrom(address token, address from, address to, uint256 amount) {
+    if (from == currentContract) {
+        // Safe require because the reference implementation would revert.
+        myBalances[token] = require_uint256(myBalances[token] - amount);
+    }
+    if (to == currentContract) {
+        // Safe require because the reference implementation would revert.
+        myBalances[token] = require_uint256(myBalances[token] + amount);
+    }
+}
+
+// Check the functional correctness of the summary of safeTransfer.
+rule checkTransferSummary(address token, address to, uint256 amount) {
+    mathint initialBalance = balanceOf(token, currentContract);
+    // Safe require because the total supply is greater than the sum of the balance of any two accounts.
+    require to != currentContract => initialBalance + balanceOf(token, to) <= to_mathint(totalSupply(token));
+
+    libSafeTransfer(token, to, amount);
```

Comment:
Note that we want this summary to be the same as the one used in `ConsistentState.spec` and `Liveness.spec`. So if we were to modify the summary to have arbitrary addresses, we would need to have a mapping with 2 entries: one for addresses, and one for token (whereas now we only have the token entry). This can be harder for the prover, but it's worth trying it out